src/MatrixTwitter.js

```diff
@@ -0,0 +1,987 @@
+            try {
+                var jsonresponse = JSON.parse(body);
+                if (jsonresponse.token_type == "bearer") {
+                    fs.writeFile("bearer.tok", jsonresponse.access_token, (err) => {
+                        if (err) {
+                            //This error is unfortunate, but not a failure to retrieve a token so the bridge can run fine.
+                            log.error("Twitter","Couldn't write bearer token to file. Reason:",err);
+                        }
+                    });
+                    //Not waiting for callback since it is trivial to get a new token, and can be done async
+                    resolve(jsonresponse.bearer_token);
+                } else {
+                    reject("Request to oauth2/post did not return the correct token type ('bearer'). This is weeeird.");
+                    console.log("Body of response:",body);
+                }
+            } catch (e) {
+                reject(e);
+            }
+        }
+    });
+  });
+}
+
+MatrixTwitter.prototype._get_bearer_token = function () {
+  return new Promise((resolve,reject) => {
+    fs.readFile('bearer.tok',{encoding:'utf-8'}, (err, content) => {
+      if(err){
+        log.warn('Twitter',"Token file not found or unreadable. Requesting new token.");
+        log.error("Twitter",err);
+        resolve(Promise.resolve(this._get_bearer_http()));
+      }
+      resolve(content);
+    });
+  }).then(token => {
+    //Test the token
+    return new Promise((resolve,reject) =>{
+      var auth = {
+        consumer_key: this.app_auth.consumer_key,
+        consumer_secret: this.app_auth.consumer_secret,
+        bearer_token: token
+      };
+      this.app_twitter = new Twitter(auth).get('application/rate_limit_status', {}, (error,status,response) => {
+        if(response.statusCode == 401){
+          log.warn('Twitter',"Authentication with existing token failed. ");
+          fs.unlink('bearer.tok', (err) => {
+            resolve(this._get_bearer_http());
+          });
+        }
+        else if (response.statusCode == 200){
+            log.info('Twitter',"Existing token OK.");
+            resolve(token);
+        }
+        else {
+            log.error("Twitter",error);
+            reject("Unexpected response to application/rate_limit_status during bearer token validation. Bailing.");
+        }
+      });
+    });
+  });
+}
+
+MatrixTwitter.prototype._get_intent = function(id){
+  return this._bridge.getIntent("@twitter_" + id + ":" + this._bridge.opts.domain);
+}
+
+MatrixTwitter.prototype._update_user_timeline_profile = function(profile){
+  var ts = new Date().getTime();
+  return this.storage.get_profile_by_id(profile.id).then((old)=>{
+    var name = true;
+    var avatar = true;
+    if(old != null){
+      name = (old.profile.name != profile.name) || (old.profile.screen_name != profile.screen_name);
+      avatar = (old.profile.profile_image_url_https != profile.profile_image_url_https);
+
```

Comment:
Variable names are misleading: I am foolishly led to believe that `name` and `avatar` are strings rather than "should I set..." (especially given `||` can be used as default assignments for exactly these kinds of use cases). Rename them to something clearer please. Secondly, is `old.profile` *always* going to exist? You `null` guard `old` but not `.profile`. Finally, whilst I appreciate you're trying to be succinct with `:170`, you're just being confusing, and may even have your check wrong as a result. Look what you're doing on `:177` - you're setting the display name to a string that uses *both* the `name` and `screen_name`. What happens if `name` is `undefined`? You're going to set the literal string "undefined" in people's display names :(